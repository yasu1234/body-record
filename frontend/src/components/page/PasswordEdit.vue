<script setup>
import axios from 'axios'
import { ref } from 'vue'
import Cookies from 'js-cookie'

import Header from '../layout/Header.vue'

const currentPassword = ref('')
const password = ref('')
const passwordConfirm = ref('')
const name = ref('')

const passwordEdit = async () => {
    try {
        const formData = new FormData();
        formData.append('current_password', currentPassword.value);
        formData.append('password', password.value);
        formData.append('password_confirmation', passwordConfirm.value);

        const res = await axios.put(import.meta.env.VITE_APP_API_BASE + `/api/v1/auth/password`, formData, {

            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        console.log({ res })
    } catch (error) {
        console.log({ error })
    }
}
</script>

<template>
    <Header />
    <h1 class="signUpTitle">パスワード変更</h1>
    <div class="singUpInput">
        <form class="form" @submit.prevent="passwordEdit">
            <div class="item">
                <label class="itemLabel" for="password">現在のパスワード</label>
                <input id="currentPassword" type="password" v-model="currentPassword">
            </div>
            <div class="item">
                <label class="itemLabel" for="password">パスワード</label>
                <input id="password" type="password" v-model="password">
            </div>
            <div class="item">
                <label class="itemLabel" for="passwordConfirm">パスワード(確認)</label>
                <input id="passwordConfirm" type="password" v-model="passwordConfirm">
            </div>
            <div class="signUpTitle">
                <button class="registerButton">更新</button>
            </div>
        </form>
    </div>
</template>

<style>
.form{
   width: 100%;
   margin:0 auto;
   padding: 50px;
   box-sizing: border-box;
}

.signUpTitle {
    padding-top: 40px;
    text-align: center;
}
 
.item{
    padding-top: 40px;
    width: 50%;
    margin: 0 auto;
}

.itemLabel{
    display: block;
    text-align: left;
}

.error-message{
    width: 50%;
    margin: 0 auto;
}
 
.error-message-text{
    color: red;
    font-size: 15px;
}

.form input[type="password"] {
     padding: 10px;
     width: 100%;
 }
 
.form textarea{
   padding: 10px;
   height: 100px;
   font-size: 14px;
   width: 50%;
 }
 
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
    padding: 10px 50px;
}
</style>
