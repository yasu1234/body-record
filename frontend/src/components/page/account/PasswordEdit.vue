<template>
  <Header />
  <TabMenu />
  <Toast position="top-center" />
  <div class="setting-container">
    <SettingSideMenu :currentIndex="3" />
    <main>
      <p class="title-line pt-10">パスワード変更</p>
      <div class="password-edit-container">
        <form class="form" @submit.prevent="checkValidate">
          <div class="pt-10">
            <label for="password">現在のパスワード</label>
            <PasswordText
              :password="currentPassword"
              :passwordType="PasswordType.password"
              @update-password="updatePassword"
            />
            <p class="validation-error-message">{{ currentPasswordError }}</p>
          </div>
          <div class="pt-10">
            <label for="password">パスワード</label>
            <PasswordText
              :password="password"
              :passwordType="PasswordType.newPassword"
              @update-password="updatePassword"
            />
            <p class="validation-error-message">{{ passwordError }}</p>
          </div>
          <div class="pt-10">
            <label for="passwordConfirm">パスワード(確認)</label>
            <PasswordText
              :password="passwordConfirm"
              :passwordType="PasswordType.newPasswordConfirm"
              @update-password="updatePassword"
            />
            <p class="validation-error-message">{{ passwordConfirmError }}</p>
          </div>
          <div class="text-center">
            <button class="password-edit-button mt-5">更新</button>
          </div>
        </form>
      </div>
    </main>
  </div>
</template>

<script setup>
import { useField, useForm } from "vee-validate";
import * as yup from "yup";
import Toast from "primevue/toast";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../../js/toast.js";
import { axiosInstance } from "../../../js/axios.js";

import Header from "../../layout/Header.vue";
import PasswordText from "../../atom/PasswordText.vue";
import SettingSideMenu from "../../layout/SettingSideMenu.vue";
import TabMenu from "../../layout/TabMenu.vue";
import { PasswordType } from "../../../js/const.js";

const toast = useToast();
const toastNotifications = new toastService(toast);

const checkValidate = async () => {
  const result = await validate();
  if (result.valid) {
    passwordEdit();
  }
};

const passwordEdit = async () => {
  try {
    const formData = new FormData();
    formData.append("current_password", currentPassword.value);
    formData.append("password", password.value);
    formData.append("password_confirmation", passwordConfirm.value);

    const res = await axiosInstance.put(`/api/v1/auth/password`, formData);
    toastNotifications.displayInfo("パスワードを変更しました", "");
  } catch (error) {
    if (error.response == null) {
      toastNotifications.displayError("パスワード変更に失敗しました", "");
      return;
    }

    let errorMessages = "";

    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      } else {
        errorMessages = error.response.data.errors;
      }
    } else if (error.response.status === 401) {
      errorMessages = "ログインしてください";
    }

    toastNotifications.displayError(
      "パスワード変更に失敗しました",
      errorMessages
    );
  }
};

const schema = yup.object({
  password: yup
    .string()
    .required("この項目は必須です")
    .min(6, "6文字以上で入力してください")
    .max(128, "128文字以下で入力してください"),
  passwordConfirm: yup
    .string()
    .required("この項目は必須です")
    .min(6, "6文字以上で入力してください")
    .max(128, "128文字以下で入力してください"),
  currentPassword: yup
    .string()
    .required("この項目は必須です")
    .min(6, "6文字以上で入力してください")
    .max(128, "128文字以下で入力してください"),
});

const { validate } = useForm({
  validationSchema: schema,
  initialValues: {
    password: "",
    passwordConfirm: "",
    currentPassword: "",
  },
});

const { value: password, errorMessage: passwordError } = useField("password");
const { value: passwordConfirm, errorMessage: passwordConfirmError } =
  useField("passwordConfirm");
const { value: currentPassword, errorMessage: currentPasswordError } =
  useField("currentPassword");

const updatePassword = (inputPassword, passwordType) => {
  if (passwordType === PasswordType.password) {
    currentPassword.value = inputPassword;
  } else if (passwordType === PasswordType.newPassword) {
    password.value = inputPassword;
  } else if (passwordType === PasswordType.newPasswordConfirm) {
    passwordConfirm.value = inputPassword;
  }
};
</script>

<style scoped>
.password-edit-container {
  width: 500px;
  margin: 0 auto;
  margin-top: 20px;
  padding: 20px;
  background-color: #ffffff;
  border: 1px solid #ccc;
  border-radius: 5px;
}
.password-edit-button {
  font-size: 16px;
  font-weight: bold;
  padding: 10px 50px;
}

@media screen and (max-width: 768px) {
  main {
    flex: 1;
  }
  .password-edit-container {
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
    padding: 20px;
    background-color: #ffffff;
    border: 1px solid #ccc;
    border-radius: 5px;
  }
}
</style>
