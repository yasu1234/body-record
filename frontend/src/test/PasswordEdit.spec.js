import { shallowMount } from '@vue/test-utils'
import PasswordEdit from '@/components/page/account/PasswordEdit.vue'

jest.mock('axios')
jest.mock('vue-router', () => {
  const router = {
    push: jest.fn(),
    pathname: '/',
    route: '',
    asPath: '',
    query: {},
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
  };
});

jest.mock("primevue/usetoast", () => ({
    useToast: jest.fn(() => ({
      add: jest.fn(),
      removeAll: jest.fn(),
    })),
  }));

jest.mock("@/js/axios.js", () => ({
  axiosInstance: {
    get: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
  setupInterceptors: jest.fn(),
}));

describe('PasswordEdit.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(PasswordEdit, {
      global: {
        stubs: ["router-link", "router-view"]
      }
    })
  })

  it('SnapShotテスト', () => {
    wrapper = shallowMount(PasswordEdit)

    expect(wrapper.element).toMatchSnapshot();
  })
})
