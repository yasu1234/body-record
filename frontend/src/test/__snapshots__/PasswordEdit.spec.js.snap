// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordEdit.vue SnapShotテスト 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <div
    class="setting-container"
  >
    <setting-side-menu-stub
      currentindex="3"
    />
    <main>
      <h1
        class="view-title pt-10"
      >
        パスワード変更
      </h1>
      <div
        class="password-edit-container"
      >
        <form
          class="form"
        >
          <div
            class="pt-10"
          >
            <label
              for="password"
            >
              現在のパスワード
            </label>
            <password-text-stub
              password=""
              passwordtype="1"
            />
            <p
              class="validation-error-message"
            />
          </div>
          <div
            class="pt-10"
          >
            <label
              for="password"
            >
              パスワード
            </label>
            <password-text-stub
              password=""
              passwordtype="3"
            />
            <p
              class="validation-error-message"
            />
          </div>
          <div
            class="pt-10"
          >
            <label
              for="passwordConfirm"
            >
              パスワード(確認)
            </label>
            <password-text-stub
              password=""
              passwordtype="4"
            />
            <p
              class="validation-error-message"
            />
          </div>
          <div
            class="text-center"
          >
            <button
              class="password-edit-button mt-5"
            >
              更新
            </button>
          </div>
        </form>
      </div>
    </main>
  </div>
  
</div>
`;
